# 📊 E-Commerce Sales Data Analysis

## 👋 Introduction
In this project, I am working with a real-world e-commerce dataset to understand sales performance and customer behavior.

This notebook focuses on:
- Understanding the dataset
- Identifying data issues
- Preparing the data for analysis and dashboarding

## 📌 Dataset Information
- Source: UCI Machine Learning Repository
- Period: December 2010 – December 2011
- Type: Transactional e-commerce data

---

## 🎯 Objectives
- Clean and prepare raw data for analysis
- Identify key sales patterns and trends
- Create features for reporting and dashboarding
- Generate business insights

---

## 🧹 Data Cleaning Plan
The dataset will be cleaned using the following steps:

1. Remove cancelled transactions (InvoiceNo starting with 'C')
2. Remove negative or invalid quantities
3. Remove invalid or zero prices
4. Handle missing values
5. Create Total Sales column (Quantity × UnitPrice)
6. Extract time-based features (Month, Year)

In [6]:
# import necessary libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np




## Step 1: Load the Dataset

In this step, I import the dataset into a Pandas DataFrame.

This allows me to:
- View the data structure
- Begin analysis
- Identify issues in the dataset

In [8]:
# Load dataset
df = pd.read_excel("../data/raw/Online Retail.xlsx")
print("Data loaded successfully.")

Data loaded successfully.


## 🔍 Step 2: Preview the Data

Here I display the first few rows of the dataset.

This helps me understand:
- What each column represents
- The kind of values in the dataset

👉 Next: I will check the dataset structure (rows, columns, data types).

In [10]:
df.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,2010-12-01 08:26:00,7.65,17850.0,United Kingdom
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,2010-12-01 08:26:00,4.25,17850.0,United Kingdom
7,536366,22633,HAND WARMER UNION JACK,6,2010-12-01 08:28:00,1.85,17850.0,United Kingdom
8,536366,22632,HAND WARMER RED POLKA DOT,6,2010-12-01 08:28:00,1.85,17850.0,United Kingdom
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,2010-12-01 08:34:00,1.69,13047.0,United Kingdom


## Step 3: Dataset Structure (df.info)

In this step, I check:
- Number of rows and columns
- Data types (text, numbers, dates)
- Missing values

This is important before cleaning the data.
 Next: I will analyze missing values in detail.

In [ ]:
# data overview
df.info()
df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


## Step 3: Dataset Structure Analysis

After loading the dataset, I examined its structure to understand the scale and quality of the data.

###  Key Observations:

- The dataset contains **541,909 rows** and **8 columns**, indicating a large transactional dataset.
- Most columns have complete data except:
  - **Description** → missing ~1,454 values
  - **CustomerID** → missing ~135,000+ values (significant)

###  Data Types:
- **Categorical (text)**: InvoiceNo, StockCode, Description, Country
- **Numeric**: Quantity, UnitPrice, CustomerID
- **Datetime**: InvoiceDate

###  Data Quality Issues Identified:
1. Missing values in `Description` and `CustomerID`
2. CustomerID has a high number of null values (this may affect customer-level analysis)
3. Dataset likely includes cancelled transactions and invalid values (to be confirmed)

 These issues must be addressed before performing analysis or building a dashboard.



##  Next Step
I will now analyze missing values in detail before starting the cleaning process.

In [5]:
# Check for missing values
missing_values = df.isnull().sum()
missing_values

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

## Step 4: Missing Values Analysis

The dataset contains missing values in two key columns:

###  Missing Values Summary:
- **Description** → 1,454 missing values  
- **CustomerID** → 135,080 missing values  

###  Observations:

1. **Description (low missing)**  
   - Only a small number of rows are missing product descriptions  
   - These rows do not provide meaningful product information  

2. **CustomerID (high missing)**  
   - A large portion of transactions (~25%) have no customer ID  
   - This means:
     - Customer-level analysis (e.g., segmentation) may be limited  
     - However, sales analysis can still proceed  



##  Cleaning Decision

- Remove rows with missing **Description** (not useful for analysis)  
- Keep rows with missing **CustomerID** for now (still useful for sales analysis)  


## ▶ Next Step
Start cleaning by removing **invalid transactions**, beginning with cancelled orders.

##  Step 5: Remove Cancelled Transactions

In this step, I remove cancelled orders from the dataset.

###  Why this is important:
- Invoice numbers starting with "C" represent cancelled transactions
- These are not actual sales and should not be included in revenue calculations
- Keeping them would lead to incorrect analysis and misleading insights

 Next: I will identify and remove all cancelled transactions.


In [12]:
# Find cancelled transactions
cancelled = df[df['InvoiceNo'].astype(str).str.startswith('C')]

cancelled.shape

(9288, 8)

In [13]:
# Remove cancelled transactions
df = df[~df['InvoiceNo'].astype(str).str.startswith('C')]

In [14]:
# Check again
df[df['InvoiceNo'].astype(str).str.startswith('C')].shape

(0, 8)

All cancelled transactions have been successfully removed.

The dataset now contains only valid sales transactions, ensuring accurate analysis going forward.

Next Step: Remove invalid values such as negative quantities.